src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -233,6 +233,16 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
             // VirtualPathData.VirtualPath returns string.Empty instead of null.
             Debug.Assert(pathData.VirtualPath != null);
 
+            // Perf: In most of the common cases, GenerateUrl is called with a null protocol, host and fragment. 
+            // In such cases, we might not need to build any Url as the url generated is mostly same as the virtual path available in pathData.
+            // For such common cases, this FastGenerateUrl method saves a string allocation per GenerateUrl call.
```

Comment:
always put `URL` in all caps in comments and documentation. @Eilon gets grumpy if you don't